luigi/contrib/redshift.py

```diff
@@ -334,16 +334,30 @@ def copy(self, cursor, f):
         If both key-based and role-based credentials are provided, role-based will be used.
         """
         logger.info("Inserting file: %s", f)
-        cursor.execute("""
-         COPY {table} from '{source}'
-         CREDENTIALS '{creds}'
-         {options}
-         ;""".format(
-            table=self.table,
-            source=f,
-            creds=self._credentials(),
-            options=self.copy_options)
-        )
+        if len(self.columns) > 0:
+            colnames = ",".join([x[0] for x in self.columns])
+            cursor.execute("""
+             COPY {table} ({colnames}) from '{source}'
+             CREDENTIALS '{creds}'
+             {options}
+             ;""".format(
+                table=self.table,
+                source=f,
+                creds=self._credentials(),
+                options=self.copy_options,
+                colnames=colnames)
+            )
+        else:
+            cursor.execute("""
+             COPY {table} from '{source}'
+             CREDENTIALS '{creds}'
+             {options}
+             ;""".format(
+                table=self.table,
+                source=f,
+                creds=self._credentials(),
+                options=self.copy_options)
+            )
```

Comment:
Perhaps you could do something more like below to reduce duplicate code

```python
colnames = ''
if len(self.columns) > 0:
    colnames = ",".join(x[0] for x in self.columns])
cursor.execute("""\
    copy {table} {cols} from '{source}'
    credentials '{creds}'
    {options}
    ;""".format(
        table=self.table,
        colnames='({})'.format(colnames),
        source=f,
        creds=self._credentials(),
        options=self.copy_options
    )
```

Reply:
Fixed :)